core/src/main/java/org/apache/accumulo/core/fate/Fate.java

```diff
@@ -397,24 +443,27 @@ public Fate(T environment, FateStore<T> store, boolean runDeadResCleaner,
       Function<Repo<T>,String> toLogStrFunc, AccumuloConfiguration conf) {
     this.store = FateLogger.wrap(store, toLogStrFunc, false);
     this.environment = environment;
-    final ThreadPoolExecutor pool = ThreadPools.getServerThreadPools().createExecutorService(conf,
+    this.transactionExecutor = ThreadPools.getServerThreadPools().createExecutorService(conf,
         Property.MANAGER_FATE_THREADPOOL_SIZE, true);
     this.workQueue = new LinkedTransferQueue<>();
+    this.runningTxRunners = new ArrayList<>();
     this.fatePoolWatcher =
         ThreadPools.getServerThreadPools().createGeneralScheduledExecutorService(conf);
     ThreadPools.watchCriticalScheduledTask(fatePoolWatcher.scheduleWithFixedDelay(() -> {
       // resize the pool if the property changed
-      ThreadPools.resizePool(pool, conf, Property.MANAGER_FATE_THREADPOOL_SIZE);
-      // If the pool grew, then ensure that there is a TransactionRunner for each thread
+      ThreadPools.resizePool(transactionExecutor, conf, Property.MANAGER_FATE_THREADPOOL_SIZE);
       final int configured = conf.getCount(Property.MANAGER_FATE_THREADPOOL_SIZE);
-      final int needed = configured - pool.getActiveCount();
+      final int needed = configured - transactionExecutor.getActiveCount();
```

Comment:
MAybe instead of using the active count could use the size of the list and do cleanup on it beforehand if the exitting was added.

```suggestion
      // remove any threads the have exited
      runningTxRunners.removeIf(tr->tr.isExited());
      final int needed = configured - runningTxRunners.size();
```

Also wondering about creating a cached thread pool (see Executors.newCachedThreadPool() ) and not bothering resizing the pool, just add and stop stuff to the pool and it will create threads as needed.  That may be too much a change for this PR though, but it would help simplify the code by removing the amount things that neede be kept in synch.

Reply:
Yes good idea.
Re the cached thread pool: this may be a good solution (I'm not sure about the implications of this yet), however would require quite a bit of refactoring, and wasn't the original intention of this PR. If that is a route we want to take, we could close this PR. This would also mean https://github.com/apache/accumulo/issues/5130 changes would be different. I assume each set of fate ops may have their own cached thread pool. Or this may make 5130 completely obsolete: all fate ops would just share one cached thread pool.

Reply:
Definitely want to keep this PR, using a cached thread pool would not solve the problems this PR is solving.  Using the cached thread pool would be complimentary to theses changes and may make the overall code simpler, but not sure.  Was thinking we could only rely on the new ` runningTxRunners` list added in this PR and not have to worry about resizing the thread pool if using a cached thread pool.  Don't worry about using it in this PR.

The cached thread pool could  simplify #5130 also as you mentioned.  I think the changes in the PR would still be built on though.

Reply:
Hmm maybe I'm misunderstanding, but wouldn't this get entirely rid of resizing of the pool making these changes obsolete?

Reply:
Right, I was misinterpreting the impact this change would have. This probably wouldn't be much refactoring. May be as simple as changing it to creating the cached pool instead of a fixed pool and deleting the `ThreadPools.resizePool(pool, conf, Property.MANAGER_FATE_THREADPOOL_SIZE);` call in the pool watcher task. In that case, it could be changed in this PR. I may be overlooking something though.

Reply:
The thing I was uncertain about w/ using a cached thread pool was how that would work w/ our utility code for creating thread pool.  If large changes are required for the utility code then would not want to make that change in the PR.  If not then it would probably be good to make the change in this PR.

Reply:
Will see about changing the thread pools util code to support this change.